File: tests/c/flogadd1.c
/* Source File: flogadd1.c */

#ifdef __CINT__
#include "cint_types.h"
#endif

#ifndef __CINT__
#include <stdio.h>
#include <stdlib.h>
#include <interface.h>
#define MAXRECORDS 1000000
#endif

/*#define MAX_SPLIT_LENGTH 256*/

int main (void)
{
  FILE *fp;
  int status, i, numrecs = 0;
  char table[TABLE_NAME_WIDTH+1]; 
  char line [1024]; /* arbitrarily long */
  char *field;
  char **spl;
  static char *ssn[MAXRECORDS];
  double salary;
  double t1, t2;

  /*
   * First, remove old .db & co. files.
   */
  if (fexists ("flogadd1.db")) {
    fileRemove ("flogadd1.db");
    fileRemove ("flogadd1.db.LCK");
    fileRemove ("f1_ssnidx.idx");
    fileRemove ("f1_ssnidx.inx");
    fileRemove ("f1_lnidx.idx");
    fileRemove ("f1_lnidx.inx");
  }
  fp = fopen ("flogadd1.df", "wb");
  if (0 == fp) {
    printf ("\n\n***Error: could not open flogadd1.df to write\n");
    return _ERROR_;
  }
  fprintf (fp, 
      "// flogadd1.df\n\n"
      "create table \"flogadd1.db\"\n"
      "  info \"Table for flogadd1.c\"\n"
      "{\n"
      "  \"LName\" char (65);\n"
      "  \"SSNumber\" char (9);\n"
      "  \"FName\" char (35);\n"
      "  \"StreetAddress\" char (75);\n"
      "  \"City\" char (50);\n"
      "  \"State\" char (2);\n"
      "  \"Zip\" char (10);\n"
      "  \"AnnualSalary\" number (10:2);\n"
      "  \"HomeOwner\" logical;\n"
      "  \"DateApplied\" date;\n"
      "  \"LastUpdated\" date;\n"
      "} indexed {\n"
      "  idx \"f1_ssnidx\" 128:case:unique \"SSNumber\";\n"
      "  idx \"f1_lnidx\" 512:case:dup \"LName\";\n"
      "};\n\n"
      );
  fclose (fp);

  status = dbcreate ("flogadd1.df");
  if (status == _ERROR_) {
    printf ("\n\n***Error: Could not dbcreate flogadd1.df: %s\n", cdberror);
    return _ERROR_;
  }
  status = remove ("flogadd1.df");
  if (status == _ERROR_) {
    printf ("\n\n***Warning: could not remove flogadd1.df\n");
  }

  field = (char *)dbopen ("flogadd1.db");
  if (0 == field) {
    printf ("\n\n***Error: Could not open flogadd1.db: %s\n", cdberror);
    return _ERROR_;
  }
  memset (table, 0, TABLE_NAME_WIDTH+1);
  strcpy (table, field);

  /*
   * OK, populate the table.
   */
  t1 = dbtime ();
  fp = fopen ("testAdd.txt", "rb");
  if (0 == fp) {
    printf ("\n\n***Error: could not open testAdd.txt to read\n");
    dbexit ();
    return _ERROR_;
  }

  for (i = 0; i < MAXRECORDS && ! feof (fp); i++) {
    if (i % 1000 == 0 && i != 0) {
      printf (".");
      fflush (stdout);
    }
    memset (line, 1020, 0);
    if (fgets (line, 1000, fp) == 0)
      break;
    if (line[0] == '\n')
      continue;
    field = strchr (line, '\n');
    if (field != 0)
      *field = 0;
    spl = split (line, ':');
    status = dbsetchar (table, "LName", spl[0]);
    if (status == _ERROR_) {
      printf ("\n\n***Error: dbsetchar %s LName %s: %s\n",
         table, spl[0], cdberror);
      dbexit ();
      return _ERROR_;
    }
    status = dbsetchar (table, "SSNumber", spl[1]);
    if (status == _ERROR_) {
      printf ("\n\n***Error: dbsetchar %s SSNumber %s: %s\n",
          table, spl[1], cdberror);
    dbexit ();
      return _ERROR_;
    }
    ssn[i] = malloc (strlen (spl[1]) + 1);
    if (0 == ssn[i]) {
      printf ("\n\n***Error: memory exhausted\n");
      dbexit();
      return _ERROR_;
    }
    strcpy (ssn[i], spl[1]);
    check_pointer (ssn[i]);
    ssn[i+1]=0;
    status = dbsetchar (table, "FName", spl[2]);
    if (status == _ERROR_) {
      printf ("\n\n***Error: dbsetchar %s FName, %s: %s\n",
	  table, spl[2], cdberror);
      dbexit();
      return _ERROR_;
    }
    status = dbsetchar (table, "StreetAddress", spl[3]);
    if (status == _ERROR_) {
      printf ("\n\n***Error: dbsetchar %s StreetAddress, %s: %s\n",
	  table, spl[3], cdberror);
      dbexit();
      return _ERROR_;
    }
    status = dbsetchar (table, "City", spl[4]);
    if (status == _ERROR_) {
      printf ("\n\n***Error: dbsetchar %s City, %s: %s\n",
	  table, spl[4], cdberror);
      dbexit();
      return _ERROR_;
    }
    status = dbsetchar (table, "State", spl[5]);
    if (status == _ERROR_) {
      printf ("\n\n***Error: dbsetchar %s State, %s: %s\n",
	  table, spl[5], cdberror);
      dbexit();
      return _ERROR_;
    }
    status = dbsetchar (table, "Zip", spl[6]);
    if (status == _ERROR_) {
      printf ("\n\n***Error: dbsetchar %s Zip, %s: %s\n",
	  table, spl[6], cdberror);
      dbexit();
      return _ERROR_;
    }
    salary = atof (spl[7]);
    if (salary < 0.00 || salary > 99999.99)
      salary = 95000.00; /* decent salary :o) */
    status = dbsetnum (table, "AnnualSalary", salary);
    if (status == _ERROR_) {
      printf ("\n\n***Error: dbsetnum %s AnnualSalary, %s: %s\n",
	  table, spl[7], cdberror);
      dbexit();
      return _ERROR_;
    }
    status = dbsetlog (table, "HomeOwner", *spl[8]);
    if (status == _ERROR_) {
      printf ("\n\n***Error: dbsetlog %s HomeOwner, %s: %s\n",
	  table, spl[8], cdberror);
      dbexit();
      return _ERROR_;
    }
    status = dbsetdate (table, "DateApplied", spl[9]);
    if (status == _ERROR_) {
      printf ("\n\n***Error: dbsetdate %s DateApplied, %s: %s\n",
	  table, spl[9], cdberror);
      dbexit();
      return _ERROR_;
    }
    status = dbsetdate (table, "LastUpdated", spl[10]);
    if (status == _ERROR_) {
      printf ("\n\n***Error: dbsetdate (%s, LastUpdated, %s): %s\n",
	  table, spl[10], cdberror);
      dbexit();
      return _ERROR_;
    }

    /*
     * Add it.
     */
    status = dbadd (table);
    if (status != _OK_) {
      if (! strcmp (cdberror, "adding record: unique index constraint violated") ) {
	printf ("\n\n***Warning: Duplicate SSN at line %d\n", i);
	continue;
      }
      printf ("\n\n***Error: calling dbadd: %s\n", cdberror);
      dbexit ();
      return _ERROR_;
    }
  }
  fclose (fp);
  numrecs = dbnumrecs (table);
  t2 = dbtime ();
  printf ("\n\nFinished adding %d records to %s in %6.4f seconds\n",
      numrecs, table, t2-t1);

  fp = fopen ("flogadd1.output", "wb");
  if (0 == fp) {
    printf ("\n\n***Error: Could not open flogadd1.output\n");
    dbexit ();
    return _ERROR_;
  }
  t1 = dbtime ();
  for (i = 0; i < MAXRECORDS && ssn[i] != 0; i++) {
    status = dbsearchexact (table, ssn[i]);
    if (status == _ERROR_) {
      printf ("\n\n***Error: dbsearch (%s, %s)\n", table, ssn[i]);
      dbexit ();
      return _ERROR_;
    }
    free (ssn[i]);
    if (status == _NOTFOUND_) {
      printf ("\n\n***Warning: didn't find %s\n", ssn[i]);
      continue;
    }
    /*
     * OK, retrieve the data and make some effort getting the data to the program.
     */
    status = dbretrieve (table);
    if (status == _ERROR_) {
      printf ("\n\n***Error: dbretrieve (%s)\n",table); 
      dbexit ();
      return _ERROR_;
    }
    /* printf ("table = %s\n", table); */
    field = dbshow (table, "LName");
    /*printf ("table = %s\n", table);*/
    if (field == 0) {
      printf ("\n\n***Error: dbshow (%s, LName): %s\n",
	  table, cdberror);
      dbexit();
      return _ERROR_;
    }
    fprintf (fp, "%s:", field);
    field = dbshow (table, "SSNumber");
    if (field == 0) {
      printf ("\n\n***Error: dbshow (%s, SSNumber): %s\n",
	  table, cdberror);
      dbexit();
      return _ERROR_;
    }
    fprintf (fp, "%s:", field);
    field = dbshow (table, "FName");
    if (field == 0) {
      printf ("\n\n***Error: dbshow (%s, FName): %s\n",
	  table, cdberror);
      dbexit();
      return _ERROR_;
    }
    fprintf (fp, "%s:", field);
    field = dbshow (table, "StreetAddress");
    if (field == 0) {
      printf ("\n\n***Error: dbshow (%s, StreetAddress): %s\n",
	  table, cdberror);
      dbexit();
      return _ERROR_;
    }
    fprintf (fp, "%s:", field);
    field = dbshow (table, "City");
    if (field == 0) {
      printf ("\n\n***Error: dbshow (%s, City): %s\n",
	  table, cdberror);
      dbexit();
      return _ERROR_;
    }
    fprintf (fp, "%s:", field);
    field = dbshow (table, "State");
    if (field == 0) {
      printf ("\n\n***Error: dbshow (%s, State): %s\n",
	  table, cdberror);
      dbexit();
      return _ERROR_;
    }
    fprintf (fp, "%s:", field);
    field = dbshow (table, "Zip");
    if (field == 0) {
      printf ("\n\n***Error: dbshow (%s, LName): %s\n",
	  table, cdberror);
      dbexit();
      return _ERROR_;
    }
    fprintf (fp, "%s:", field);
    field = dbshow (table, "AnnualSalary");
    if (field == 0) {
      printf ("\n\n***Error: dbshow (%s, AnnualSalary): %s\n",
	  table, cdberror);
      dbexit();
      return _ERROR_;
    }
    fprintf (fp, "%s:", field);
    field = dbshow (table, "HomeOwner");
    if (field == 0) {
      printf ("\n\n***Error: dbshow (%s, HomeOwner): %s\n",
	  table, cdberror);
      dbexit();
      return _ERROR_;
    }
    fprintf (fp, "%s:", field);
    field = dbshow (table, "DateApplied");
    if (field == 0) {
      printf ("\n\n***Error: dbshow (%s, DateApplied): %s\n",
	  table, cdberror);
      dbexit();
      return _ERROR_;
    }
    fprintf (fp, "%s:", field);
    field = dbshow (table, "LastUpdated");
    if (field == 0) {
      printf ("\n\n***Error: dbshow (%s, LastUpdated): %s\n",
	  table, cdberror);
      dbexit();
      return _ERROR_;
    }
    fprintf (fp, "%s\n", field);
  }
  t2 = dbtime ();
  fclose (fp);
  printf ("\n\nFinished searching %d records to %s in %6.4f seconds\n",
      numrecs, table, t2-t1);

  /* Done */
  dbexit ();
  print_block_list();
  return _OK_;
}

File: tests/python/flogadd1.py
#!/usr/bin/python2.5

# Load the python module.
import sys
import os
sys.path[0] = ''
os.chdir ("../../lib")
import cdbpy
os.chdir ("../tests/python")

# Lock and load.
fadd1 = open ("flogadd1.df", "w", 0644)
file.write (fadd1, "// flogadd1.df\n")
file.write (fadd1, 'create table "flogadd1.db"\n')
file.write (fadd1, '  info "Table for flogadd1.py"\n')
file.write (fadd1, "{\n")
file.write (fadd1, '  "LName" char (65);\n')
file.write (fadd1, '  "SSNumber" char (9);\n')
file.write (fadd1, '  "FName" char (35);\n')
file.write (fadd1, '  "StreetAddress" char (80);\n')
file.write (fadd1, '  "City" char (50);\n')
file.write (fadd1, '  "State" char (2);\n')
file.write (fadd1, '  "Zip" char (10);\n')
file.write (fadd1, '  "AnnualSalary" number (10:2);\n')
file.write (fadd1, '  "HomeOwner" logical;\n')
file.write (fadd1, '  "DateApplied" date;\n')
file.write (fadd1, '  "LastUpdated" time;\n')
file.write (fadd1, '} indexed {\n')
file.write (fadd1, '  idx "f1_ssnidx" 256:case:unique "SSNumber";\n')
file.write (fadd1, '  idx "f1_lnidx" 512:case:dup "LName";\n')
file.write (fadd1, '};')
file.write (fadd1, "\n\n")

file.close (fadd1)

if (os.access ("flogadd1.db", os.F_OK)) :
  os.remove ("flogadd1.db")

if (os.access ("flogadd1.db.LCK", os.F_OK)) :
  os.remove ("flogadd1.db.LCK")

if (os.access ("f1_ssnidx.inx", os.F_OK)) :
  os.remove ("f1_ssnidx.inx")

if (os.access ("f1_ssnidx.idx", os.F_OK)) :
  os.remove ("f1_ssnidx.idx")

if (os.access ("f1_lnidx.inx", os.F_OK)) :
  os.remove ("f1_lnidx.inx")

if (os.access ("f1_lnidx.idx", os.F_OK)) :
  os.remove ("f1_lnidx.idx")

#
# Create the table and indexes
#
status = cdbpy.dbcreate ("flogadd1.df")
if (status != 0) :
  print "%s %s" % ("\n\n***Error: dbcreate flogadd1.df:", cdbpy.dbgeterror())
  sys.exit (1)

os.remove ("flogadd1.df")

table = cdbpy.dbopen ("flogadd1.db")
if (table == '') :
  print "%s %s" % ("\n\n***Error: dbopen flogadd1.db:", cdbpy.dbgeterror())
  sys.exit (1)

print ("Parsing testAdd.txt")

s = open ("testAdd.txt").read()
a1 = []
a1 = s.split()

dupList = []

t1 = 0.0
t1 = cdbpy.dbtime ()
counter = 0
for ln in a1 :
  counter += 1
  if (counter % 1000 == 0) :
    sys.stdout.write('.')
    sys.stdout.flush()
  a2 = []
  a2 = ln.split(':')
  status = cdbpy.dbsetchar (table, "LName", a2[0])
  if (status == -1) :
    print "%s %s:%s" % \
        ("\n\n***Error: dbsetchar table LName ", a2[0], cdbpy.dbgeterror())
    cdbpy.dbclose (table)
    sys.exit (1)
  status = cdbpy.dbsetchar (table, "SSNumber", a2[1])
  if (status == -1) :
    print "%s %s:%s" % \
        ("\n\n***Error: dbsetchar table SSNumber ", a2[1], cdbpy.dbgeterror())
    cdbpy.dbclose (table)
    sys.exit (1)
  status = cdbpy.dbsetchar (table, "FName", a2[2])
  if (status == -1) :
    print "%s %s:%s" % \
        ("\n\n***Error: dbsetchar table FName ", a2[2], cdbpy.dbgeterror())
    cdbpy.dbclose (table)
    sys.exit (1)
  status = cdbpy.dbsetchar (table, "StreetAddress", a2[3])
  if (status == -1) :
    print "%s %s:%s" % \
        ("\n\n***Error: dbsetchar table StreetAddress ", a2[3], cdbpy.dbgeterror())
    cdbpy.dbclose (table)
    sys.exit (1)
  status = cdbpy.dbsetchar (table, "City", a2[4])
  if (status == -1) :
    print "%s %s:%s" % \
        ("\n\n***Error: dbsetchar table City ", a2[4], cdbpy.dbgeterror())
    cdbpy.dbclose (table)
    sys.exit (1)
  status = cdbpy.dbsetchar (table, "State", a2[5])
  if (status == -1) :
    print "%s %s:%s" % \
        ("\n\n***Error: dbsetchar table State ", a2[5], cdbpy.dbgeterror())
    cdbpy.dbclose (table)
    sys.exit (1)
  status = cdbpy.dbsetchar (table, "Zip", a2[6])
  if (status == -1) :
    print "%s %s:%s" % \
        ("\n\n***Error: dbsetchar table Zip ", a2[6], cdbpy.dbgeterror())
    cdbpy.dbclose (table)
    sys.exit (1)
  status = cdbpy.dbsetnum (table, "AnnualSalary", float(a2[7]))
  if (status == -1) :
    print "%s %s:%s" % \
        ("\n\n***Error: dbsetchar table AnnualSalary ", a2[7], cdbpy.dbgeterror())
    cdbpy.dbclose (table)
    sys.exit (1)
  status = cdbpy.dbsetlog (table, "HomeOwner", a2[8])
  if (status == -1) :
    print "%s %s:%s" % \
        ("\n\n***Error: dbsetchar table HomeOwner ", a2[8], cdbpy.dbgeterror())
    cdbpy.dbclose (table)
    sys.exit (1)
  status = cdbpy.dbsetdate (table, "DateApplied", "0")
  if (status == -1) :
    print "%s %s:%s" % \
        ("\n\n***Error: dbsetchar table DateApplied ", 0, cdbpy.dbgeterror())
    cdbpy.dbclose (table)
    sys.exit (1)
  status = cdbpy.dbsettime (table, "LastUpdated", "0")
  if (status == -1) :
    print "%s %s:%s" % \
        ("\n\n***Error: dbsetchar table FName ", 0, cdbpy.dbgeterror())
    cdbpy.dbclose (table)
    sys.exit (1)
  status = cdbpy.dbadd (table)
  if (status == -1) :
    terr = cdbpy.dbgeterror()
    status = cmp (terr, "adding record: unique index constraint violated")
    if not status :
      print "***Warning: Duplicate SSN at line %d" % (counter)
      dupList.append (a2[1])
      continue
    print "%s:%s" % \
        ("\n\n***Error: dbadd table ", cdbpy.dbgeterror())
    cdbpy.dbclose (table)
    sys.exit (1)
  del a2

sys.stdout.write ("\n")
t2 = 0.0
t2 = cdbpy.dbtime ()
num = cdbpy.dbnumrecs (table)

print "Finished adding to %s: %d in %6.4f seconds" % (table, num, t2-t1)

#
# Now, traverse through the table and insure that the data is consistent with
# what was actually entered.
#

t1 = 0.0
t1 = cdbpy.dbtime ()
counter = 0
cdbpy.dbcurrent (table, "f1_ssnidx")
status = cdbpy.dbiseof (table)
if (status == 1) :
  print "There are no records in %s" % (table)
  cdbpy.dbexit ()
  sys.exit (1)

for ln in a1 :
  counter += 1
  if (counter % 1000 == 0) :
    sys.stdout.write('.')
    sys.stdout.flush()
  a2 = []
  a2 = ln.split(':')
  status = cdbpy.dbsearchexact (table, a2[1])
  if (status == 0) :
    print "***Warning: couldn't find %s in the database: line %d" % (a2[0], counter)
    continue
  status = cdbpy.dbretrieve (table)
  if (status == -1) :
    print "\n\n***Error: Could not retrieve from %s: %s" % (table, cdbpy.dbgeterror())
    cdbpy.dbexit ()
    sys.exit (1)
  header = 0
  LName = cdbpy.dbshow (table, "LName")
  status = cmp (LName, a2[0])
  if (status != 0) :
    if (a2[1] in dupList): continue
    if not header :
      print "***Warning: Record out of sync ",
      header = 1
    print "LName",
  SSNumber = cdbpy.dbshow (table, "SSNumber")
  status = cmp (SSNumber, a2[1])
  if (status != 0) :
    if (a2[1] in dupList): continue
    if not header :
      print "***Warning: Record out of sync ",
      header = 1
    print "SSNumber",
  FName = cdbpy.dbshow (table, "FName")
  status = cmp (FName, a2[2])
  if (status != 0) :
    if (a2[1] in dupList): continue
    if not header :
      print "***Warning: Record out of sync ",
      header = 1
    print "FName",
  StreetAddress = cdbpy.dbshow (table, "StreetAddress")
  status = cmp (StreetAddress, a2[3])
  if (status != 0) :
    if (a2[1] in dupList): continue
    if not header :
      print "***Warning: Record out of sync ",
      header = 1
    print "StreetAddress",
  City = cdbpy.dbshow (table, "City")
  status = cmp (City, a2[4])
  if (status != 0) :
    if (a2[1] in dupList): continue
    if not header :
      print "***Warning: Record out of sync ",
      header = 1
    print "City",
  Zip = cdbpy.dbshow (table, "Zip")
  status = cmp (Zip, a2[6])
  if (status != 0) :
    if (a2[1] in dupList): continue
    if not header :
      print "***Warning: Record out of sync ",
      header = 1
    print "Zip",
  AnnualSalary = cdbpy.dbshow (table, "AnnualSalary")
  status = cmp (AnnualSalary, a2[7])
  if (status != 0) :
    if (a2[1] in dupList): continue
    if not header :
      print "***Warning: Record out of sync ",
      header = 1
    print "AnnualSalary",
  del a2

del dupList

t2 = 0.0
t2 = cdbpy.dbtime ()
num = cdbpy.dbnumrecs (table)

print "\nFinished verifying to %s: %d in %6.4f seconds" % (table, num, t2-t1)

cdbpy.dbclose (table)

sys.exit (0)
